### PR TITLE
Enforce string limits for deserialization

### PR DESCRIPTION
**Pull request by zbuc**
## Describe your changes

This follows the suggestion from @hdevalence in https://github.com/penumbra-zone/penumbra/issues/3620#issuecomment-2141013858 to add validation of user-supplied `String` lengths to the various `TryFrom` protobuf deserialization implementations.

The user-supplied `String` values concerned are found within the governance and staking crates, concerning governance proposals and validator definitions.

**This is consensus breaking** as historic data **may not comply with the new limitations**. Thus there may be unexpected panics when trying to retrieve and deserialize data from storage.

An additional migration has been included to truncate any existing data that does not adhere to the new limits.

The strings are configured to use **byte length limits** rather than **character length limits**. This seems favorable because values will be guaranteed to fit into fixed-size buffers of the byte length limit for the field. Code based on https://github.com/rust-lang/rust/issues/93743 was pulled in to perform the truncation at UTF-8 character boundaries.

## Issue ticket number and link

Closes https://github.com/penumbra-zone/penumbra/issues/3620

## Checklist before requesting a review

- [x] If this code contains consensus-breaking changes, I have added the "consensus-breaking" label. Otherwise, I declare my belief that there are not consensus-breaking changes, for the following reason:

### CODE REVIEW

File: crates/core/component/governance/src/proposal.rs [hunk 1]
```diff
@@ -82,6 +82,17 @@ impl TryFrom<pb::Proposal> for Proposal {
...
 
     fn try_from(inner: pb::Proposal) -> Result<Proposal, Self::Error> {
+        // Validation (matches limits from `impl AppActionHandler for ProposalSubmit`):
+        // - Title has a max length of 80 chars
+        if inner.title.len() > 80 {
+            anyhow::bail!("proposal title field must be less than 80 characters");
+        }
+
+        // - Description has a max length of 10_000 chars
+        if inner.description.len() > 10_000 {
```
Comment (erwanor) [inline crates/core/component/governance/src/proposal.rs:92]: This makes sense, CSDK has a similar limit.

File: crates/core/component/governance/src/proposal.rs [hunk 2]
```diff
@@ -82,6 +82,17 @@ impl TryFrom<pb::Proposal> for Proposal {
     type Error = anyhow::Error;
 
     fn try_from(inner: pb::Proposal) -> Result<Proposal, Self::Error> {
+        // Validation (matches limits from `impl AppActionHandler for ProposalSubmit`):
+        // - Title has a max length of 80 chars
+        if inner.title.len() > 80 {
```
Comment (erwanor) [inline crates/core/component/governance/src/proposal.rs:87]: CSDK limit is 255 characters. This works too imo.

File: crates/core/component/governance/src/proposal.rs [hunk 3]
```diff
@@ -95,6 +106,11 @@ impl TryFrom<pb::Proposal> for Proposal {
                     commit: if signaling.commit.is_empty() {
                         None
                     } else {
+                        // Commit hash has max length of 64 chars:
+                        if signaling.commit.len() > 64 {
```
Comment (erwanor) [inline crates/core/component/governance/src/proposal.rs:110]: How about we make this 255 so it can be used with IPFS links, fine to leave as is too

File: crates/core/component/governance/src/proposal.rs [hunk 4]
```diff
@@ -123,9 +139,15 @@ impl TryFrom<pb::Proposal> for Proposal {
                 Payload::UpgradePlan(upgrade_plan) => ProposalPayload::UpgradePlan {
                     height: upgrade_plan.height,
                 },
-                Payload::FreezeIbcClient(freeze_ibc_client) => ProposalPayload::FreezeIbcClient {
-                    client_id: freeze_ibc_client.client_id,
-                },
+                Payload::FreezeIbcClient(freeze_ibc_client) => {
+                    // Validation: client ID has a max length of 128 chars
+                    if freeze_ibc_client.client_id.len() > 128 {
```
Comment (erwanor) [inline crates/core/component/governance/src/proposal.rs:144]: Makes sense to me

File: crates/core/component/governance/src/proposal_state.rs [hunk 5]
```diff
@@ -292,6 +292,11 @@ impl TryFrom<pb::ProposalOutcome> for Outcome<String> {
                     withdrawn,
                 }) => Outcome::Failed {
                     withdrawn: if let Some(pb::proposal_outcome::Withdrawn { reason }) = withdrawn {
+                        // Max reason length is 1kb
+                        if reason.len() > 1024 {
+                            anyhow::bail!("withdrawn reason is too long")
```
Comment (erwanor) [inline crates/core/component/governance/src/proposal_state.rs:297]: Makes sense, i'd put the charlimit in the error message so this bubble up as a useful message

File: crates/core/component/stake/src/validator.rs [hunk 6]
```diff
@@ -133,6 +133,22 @@ impl TryFrom<ValidatorToml> for Validator {
...
+
+        // - Name has a max length of 140 chars
+        if v.name.len() > 140 {
+            anyhow::bail!("validator name must be less than 140 characters");
+        }
+
+        // - Description has a max length of 280 chars
+        if v.description.len() > 280 {
+            anyhow::bail!("validator description must be less than 280 characters");
+        }
```
Comment (erwanor) [inline crates/core/component/stake/src/validator.rs:150]: ACK. The funding stream check is in the `FundingStreams` conversion impl.

File: crates/core/component/stake/src/validator/definition.rs [hunk 7]
```diff
@@ -29,11 +29,27 @@ impl From<Definition> for pb::ValidatorDefinition {
...
+        // - Name has a max length of 140 chars
+        if validator.name.len() > 140 {
+            anyhow::bail!("validator name must be less than 140 characters");
+        }
+
+        // - Description has a max length of 280 chars
+        if validator.description.len() > 280 {
+            anyhow::bail!("validator description must be less than 280 characters");
+        }
+
```
Comment (erwanor) [inline crates/core/component/stake/src/validator/definition.rs:50]: I don't think we need/want this because we construct a `Validator` down below which is where the validation occurs

File: crates/bin/pd/src/migrate.rs [hunk 8]
```diff
@@ -72,37 +80,96 @@ impl Migration {
...
+            .await
+            .expect("chain state has a block height");
+        let post_upgrade_height = pre_upgrade_height.wrapping_add(1);
+
+        let migration_duration = match self {
+            Migration::SimpleMigration => simple::migrate(storage.clone()).await?,
+
+            Migration::Testnet78 => testnet78::migrate(storage.clone()).await?,
+            _ => unreachable!(),
+        };
```
Comment (erwanor) [inline crates/bin/pd/src/migrate.rs:107]: I don't think we should do this specific refactor, it doesn't address the module's main issues (brittleness) but it shuffles important code paths around, right after we tested them, but ahead of having integration tests ready (#4337).

In its current state, it makes genesis generation work against a stale snapshot which is a bug. It would be caught by integration tests but this is an example of how deliberate we have to be about reworking this, especially now.

**Review by erwanor — CHANGES_REQUESTED** [review]
Relative to the proposed migration structure, the limits lgtm

File: crates/bin/pd/src/migrate.rs (hunk 8 above)
Comment (zbuc) [reply crates/bin/pd/src/migrate.rs:107]: I'm mixed on whether to refactor right now or not. I hear you about the testing, however adding the new migration seems to require a careful copy-pasting of genesis handling code and manual testing anyways. I like your suggestion of moving everything into a `Trait` impl.

File: crates/core/component/stake/src/validator/definition.rs (hunk 7 above)
Comment (zbuc) [reply crates/core/component/stake/src/validator/definition.rs:50]: Since the `Definition` can get included on-chain (as part of `Action::ValidatorDefinition`) it seems important to apply a length limit to it.

File: crates/bin/pd/src/migrate.rs (hunk 8 above)
Comment (zbuc) [reply crates/bin/pd/src/migrate.rs]: Reverted the refactoring for now

File: crates/core/component/governance/src/validator_vote/action.rs [hunk 9]
```diff
@@ -4,7 +4,7 @@ use penumbra_stake::{GovernanceKey, IdentityKey};
 use penumbra_txhash::{EffectHash, EffectingData};
 use serde::{Deserialize, Serialize};
 
-use crate::vote::Vote;
+use crate::{vote::Vote, MAX_VALIDATOR_VOTE_REASON_LENGTH};
```
Comment (cronokirby) [inline crates/core/component/governance/src/validator_vote/action.rs:7]: Maybe we should use this constant in proposal_state.rs too? If it requires changing visibility then that's not worth it, just wondering.

File: crates/bin/pd/src/migrate/testnet72.rs [hunk 10]
```diff
@@ -1,5 +1,5 @@
 //! Contains functions related to the migration script of Testnet72
-
+#![allow(dead_code)]
```
Comment (cronokirby) [inline crates/bin/pd/src/migrate/testnet72.rs:2]: Why was this added to the old migrations?

File: crates/bin/pd/src/migrate/testnet78.rs [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+    if index >= s.len() {
+        s.len()
+    } else {
+        let lower_bound = index.saturating_sub(3);
+        let new_index = s.as_bytes()[lower_bound..=index]
+            .iter()
+            .rposition(|b| is_utf8_char_boundary(*b));
+
+        // SAFETY: we know that the character boundary will be within four bytes
+        unsafe { lower_bound + new_index.unwrap_unchecked() }
```
Comment (cronokirby) [inline crates/bin/pd/src/migrate/testnet78.rs:357]:
```suggestion
        lower_bound + new_index.unwrap()
```

Why don't we just avoid the unsafe thing altogether? Seems like we'd prefer a panic in migration testing, if for some reason this code is mistaken (this does seem safe to me, but I don't see the benefit in using an unchecked method)

**Review by cronokirby — APPROVED** [review]

File: crates/bin/pd/src/migrate/testnet72.rs (hunk 10 above)
Comment (zbuc) [reply crates/bin/pd/src/migrate/testnet72.rs:2]: They are no longer callable via `pcli` (since they are only valid for the penumbra version contemporary to the migration script) but we wanted to keep the code around to serve as an example of how to write a migration.
